docs/sbft-formal-model/replica.i.dfy

```diff
@@ -473,14 +525,52 @@ module Replica {
     && v' == v.(newViewMsgsRecvd := v.newViewMsgsRecvd.(msgs := v.newViewMsgsRecvd.msgs + {msg}))
   }
 
+  predicate SendCheckpoint(c:Constants, v:Variables, v':Variables, msgOps:Network.MessageOps<Message>, seqID:SequenceID) {
+    && v.WF(c)
+    && msgOps.IsSend()
+    && var msg := msgOps.send.value;
+    && msg.payload.CheckpointMsg?
+    && msg.payload.seqIDReached <= v.countExecutedSeqIDs
+    && v == v'
+  }
+
+  predicate RecvCheckpoint(c:Constants, v:Variables, v':Variables, msgOps:Network.MessageOps<Message>) {
+    && v.WF(c)
+    && msgOps.IsRecv()
+    && var msg := msgOps.recv.value;
+    && msg.payload.CheckpointMsg?
+    && v' == v.(checkpointMsgsRecvd := v.checkpointMsgsRecvd.(msgs := v.checkpointMsgsRecvd.msgs + {msg}))
+  }
+
+  predicate HasStableCheckpointForSeqID(c:Constants, v:Variables, seqID:SequenceID) {
+    && v.WF(c)
+    && var relevantCheckpointMsgs := set msg | && msg in v.checkpointMsgsRecvd.msgs 
+                                               && msg.payload.CheckpointMsg?
+                                               && msg.payload.seqIDReached == seqID;
+    && |relevantCheckpointMsgs| >= c.clusterConfig.AgreementQuorum()
+  }
+
+  predicate AdvanceWorkingWindow(c:Constants, v:Variables, v':Variables, msgOps:Network.MessageOps<Message>, seqID:SequenceID) {
+    && v.WF(c)
+    && msgOps.NoSendRecv()
+    && seqID > v.workingWindow.lastStableCheckpoint
```

Comment:
Shouldn't the pre-condition be stricter, and say something like `seqID > ls + win`, like in the paper - where `win` is the size of the active window?

Reply:
Ohh I see, good point.